envoy/config/filter/http/jwt_authn/v2/config.proto

```diff
@@ -0,0 +1,207 @@
+  // See `OpenID Discovery
+  // <https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata>`_
+  //
+  // Example:  jwks_uri:
+  //
+  // .. code-block:: yaml
+  //
+  //    jwks_uri:
+  //    - uri: https://www.googleapis.com/oauth2/v1/certs
+  //      cluster: jwt.www.googleapis.com|443
+  //
+  envoy.api.v2.core.HttpUri jwks_uri = 3;
+
+  // Duration after which the cached public key should be expired. If not
+  // specified, default cache duration is 5 minutes.
+  google.protobuf.Duration public_key_cache_duration = 4;
+
+  // If false, the JWT is removed in the request after a success verification.
+  // Set it true, if don't want it to be removed.
+  // Default value is false which is to remove JWT.
+  bool forward_jwt = 5;
+
+  // If true, the request is allowed if the JWT verification fails.
+  // Default value is false; JWT verification failed request will be rejected.
+  bool allow_failed_jwt = 6;
+
+  // Two fields below define where to extract the JWT from an HTTP request.
+  //
+  // If no explicit location is specified, the following default locations are
+  // tried in order:
+  //
+  // 1. The Authorization header using the Bearer schema.
+  //    See `here <https://tools.ietf.org/html/rfc6750#section-2.1>`_
+  //    Example::
+  //
+  //       Authorization: Bearer <token>.
+  //
+  // 2. `access_token` query parameter.
+  //    See `here <https://tools.ietf.org/html/rfc6750#section-2.3>`_
+  //
+  //
+  // Only the first JWT token is verified per request if multiple tokens are
+  // specified.
+  //
+  // If both jwt_headers and jwt_params are specified, jwt_headers will be
+  // checked first.
+  //
+
+  // Specify the HTTP headers to extract JWT token.
+  //
+  // For examples, following config::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_headers:
+  //   - name: x-goog-iap-jwt-assertion
+  //
+  // can be used to extract token from header::
+  //
+  //   x-goog-iap-jwt-assertion: <JWT>.
+  //
+  repeated JwtHeader jwt_headers = 7;
+
+  // JWT is sent in a query parameter. `jwt_params` represents the
+  // query parameter names.
+  //
+  // For example, if config is::
+  //
+  // .. code-block:: yaml
+  //
+  //   jwt_params:
+  //   - jwt_token
+  //
+  // The query parameter is::
+  //
+  //    /path?jwt_token=<JWT>
+  //
+  repeated string jwt_params = 8;
+
+  // Following two fields specify if the payload of a successfully verified
+  // JWT should be forwarded to the backend, and how to forward it.
+  // The forwarded data is: base64_encoded(jwt_payload_in_JSON)
+
+  // If false, JWT payload will be forwarded. This is the default value.
+  // If true, it is not forwarded.
+  bool not_jwt_payload = 9;
+
+  // This field specifies the header name to forward JWT payload.
+  // If it is not specified, the name "sec-istio-auth-userinfo" will be used.
+  // If multiple JWTs needed to be forwarded, please use different header names.
+  string jwt_payload_header = 10;
```

Comment:
Would it be cleaner to have only this filed, without not_jwt_payload boolean. Empty string means payload is not forwarded. The default behavior will be NOT sending payload, but I think it's better than send it to some 'default' location that might not work for everyone.